// langium/lib/grammar/langium-grammar-validator.js
"use strict";
/******************************************************************************
 * Copyright 2021 TypeFox GmbH
 * This program and the accompanying materials are made available under the
 * terms of the MIT License, which is available in the project root.
 ******************************************************************************/
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LangiumGrammarValidator = exports.IssueCodes = exports.LangiumGrammarValidationRegistry = void 0;
const path_1 = __importDefault(require("path"));
const vscode_languageserver_types_1 = require("vscode-languageserver-types");
const vscode_uri_1 = require("vscode-uri");
const ast_util_1 = require("../utils/ast-util");
const collections_1 = require("../utils/collections");
const cst_util_1 = require("../utils/cst-util");
const stream_1 = require("../utils/stream");
const validation_registry_1 = require("../validation/validation-registry");
const ast = __importStar(require("./generated/ast"));
const grammar_util_1 = require("./grammar-util");
const type_validator_1 = require("./type-system/type-validator");
class LangiumGrammarValidationRegistry extends validation_registry_1.ValidationRegistry {
    constructor(services) {
        super(services);
        const validator = services.validation.LangiumGrammarValidator;
        const checks = {
            AbstractRule: validator.checkRuleName,
            ParserRule: [
                validator.checkParserRuleDataType,
                validator.checkRuleParametersUsed
            ],
            TerminalRule: [
                validator.checkTerminalRuleReturnType,
                validator.checkHiddenTerminalRule,
                validator.checkEmptyTerminalRule
            ],
            Keyword: validator.checkKeyword,
            UnorderedGroup: validator.checkUnorderedGroup,
            Grammar: [
                validator.checkGrammarName,
                validator.checkEntryGrammarRule,
                validator.checkUniqueRuleName,
                validator.checkUniqueTypeName,
                validator.checkUniqueImportedRules,
                validator.checkDuplicateImportedGrammar,
                validator.checkGrammarHiddenTokens,
                validator.checkGrammarForUnusedRules,
                validator.checkGrammarImports,
                validator.checkGrammarTypeAliases,
                validator.checkGrammarTypeInfer,
                validator.checkTypesConsistency,
                validator.checkPropertyNameDuplication
            ],
            GrammarImport: validator.checkPackageImport,
            CharacterRange: validator.checkInvalidCharacterRange,
            RuleCall: [
                validator.checkUsedHiddenTerminalRule,
                validator.checkUsedFragmentTerminalRule,
                validator.checkRuleCallParameters,
            ],
            TerminalRuleCall: validator.checkUsedHiddenTerminalRule,
            CrossReference: [
                validator.checkCrossReferenceSyntax,
                validator.checkCrossRefNameAssignment,
                validator.checkCrossRefTerminalType,
                validator.checkCrossRefType
            ],
            AtomType: [
                validator.checkAtomTypeRefType
            ]
        };
        this.register(checks, validator);
    }
}
exports.LangiumGrammarValidationRegistry = LangiumGrammarValidationRegistry;
var IssueCodes;
(function (IssueCodes) {
    IssueCodes.GrammarNameUppercase = 'grammar-name-uppercase';
    IssueCodes.RuleNameUppercase = 'rule-name-uppercase';
    IssueCodes.HiddenGrammarTokens = 'hidden-grammar-tokens';
    IssueCodes.UseRegexTokens = 'use-regex-tokens';
    IssueCodes.EntryRuleTokenSyntax = 'entry-rule-token-syntax';
    IssueCodes.CrossRefTokenSyntax = 'cross-ref-token-syntax';
    IssueCodes.MissingImport = 'missing-import';
    IssueCodes.UnnecessaryFileExtension = 'unnecessary-file-extension';
    IssueCodes.InvalidReturns = 'invalid-returns';
    IssueCodes.InvalidInfers = 'invalid-infers';
    IssueCodes.MissingInfer = 'missing-infer';
    IssueCodes.SuperfluousInfer = 'superfluous-infer';
})(IssueCodes = exports.IssueCodes || (exports.IssueCodes = {}));
class LangiumGrammarValidator {
    constructor(services) {
        this.references = services.references.References;
        this.documents = services.shared.workspace.LangiumDocuments;
    }
    checkGrammarName(grammar, accept) {
        if (grammar.name) {
            const firstChar = grammar.name.substring(0, 1);
            if (firstChar.toUpperCase() !== firstChar) {
                accept('warning', 'Grammar name should start with an upper case letter.', { node: grammar, property: 'name', code: IssueCodes.GrammarNameUppercase });
            }
        }
    }
    checkEntryGrammarRule(grammar, accept) {
        const entryRules = grammar.rules.filter(e => ast.isParserRule(e) && e.entry);
        if (grammar.isDeclared && entryRules.length === 0) {
            const possibleEntryRule = grammar.rules.find(e => ast.isParserRule(e) && !(0, grammar_util_1.isDataTypeRule)(e));
            if (possibleEntryRule) {
                accept('error', 'The grammar is missing an entry parser rule. This rule can be an entry one.', { node: possibleEntryRule, property: 'name', code: IssueCodes.EntryRuleTokenSyntax });
            }
            else {
                accept('error', 'This grammar is missing an entry parser rule.', { node: grammar, property: 'name' });
            }
        }
        else if (!grammar.isDeclared && entryRules.length >= 1) {
            entryRules.forEach(rule => accept('error', 'Cannot declare entry rules for unnamed grammars.', { node: rule, property: 'name' }));
        }
        else if (entryRules.length > 1) {
            entryRules.forEach(rule => accept('error', 'The entry rule has to be unique.', { node: rule, property: 'name' }));
        }
        else if ((0, grammar_util_1.isDataTypeRule)(entryRules[0])) {
            accept('error', 'The entry rule cannot be a data type rule.', { node: entryRules[0], property: 'name' });
        }
    }
    /**
     * Check whether any rule defined in this grammar is a duplicate of an already defined rule or an imported rule
     */
    checkUniqueRuleName(grammar, accept) {
        this.checkUniqueName(grammar, accept, (grammar) => grammar.rules, 'rule');
    }
    /**
     * Check whether any type defined in this grammar is a duplicate of an already defined type or an imported type
     */
    checkUniqueTypeName(grammar, accept) {
        this.checkUniqueName(grammar, accept, (grammar) => grammar.types.concat(grammar.interfaces), 'type');
    }
    checkUniqueName(grammar, accept, extractor, uniqueObjName) {
        const map = new collections_1.MultiMap();
        extractor(grammar).forEach(e => map.add(e.name, e));
        for (const name of map.keys()) {
            const types = map.get(name);
            if (types.length > 1) {
                types.forEach(e => {
                    accept('error', `A ${uniqueObjName}'s name has to be unique.`, { node: e, property: 'name' });
                });
            }
        }
        const imported = new Set();
        const resolvedGrammars = (0, grammar_util_1.resolveTransitiveImports)(this.documents, grammar);
        for (const resolvedGrammar of resolvedGrammars) {
            extractor(resolvedGrammar).forEach(e => imported.add(e.name));
        }
        for (const name of map.keys()) {
            if (imported.has(name)) {
                const types = map.get(name);
                types.forEach(e => {
                    accept('error', `A ${uniqueObjName} with the name '${e.name}' already exists in an imported grammar.`, { node: e, property: 'name' });
                });
            }
        }
    }
    checkDuplicateImportedGrammar(grammar, accept) {
        const importMap = new collections_1.MultiMap();
        for (const imp of grammar.imports) {
            const resolvedGrammar = (0, grammar_util_1.resolveImport)(this.documents, imp);
            if (resolvedGrammar) {
                importMap.add(resolvedGrammar, imp);
            }
        }
        for (const grammar of importMap.keys()) {
            const imports = importMap.get(grammar);
            if (imports.length > 1) {
                imports.forEach((imp, i) => {
                    if (i > 0) {
                        accept('warning', 'The grammar is already being directly imported.', { node: imp, tags: [vscode_languageserver_types_1.DiagnosticTag.Unnecessary] });
                    }
                });
            }
        }
    }
    /**
     * Compared to the validation above, this validation only checks whether two imported grammars export the same grammar rule.
     */
    checkUniqueImportedRules(grammar, accept) {
        const imports = new Map();
        for (const imp of grammar.imports) {
            const importedGrammars = (0, grammar_util_1.resolveTransitiveImports)(this.documents, imp);
            imports.set(imp, importedGrammars);
        }
        const allDuplicates = new collections_1.MultiMap();
        for (const outerImport of grammar.imports) {
            const outerGrammars = imports.get(outerImport);
            for (const innerImport of grammar.imports) {
                if (outerImport === innerImport) {
                    continue;
                }
                const innerGrammars = imports.get(innerImport);
                const duplicates = this.getDuplicateExportedRules(outerGrammars, innerGrammars);
                for (const duplicate of duplicates) {
                    allDuplicates.add(outerImport, duplicate);
                }
            }
        }
        for (const imp of grammar.imports) {
            const duplicates = allDuplicates.get(imp);
            if (duplicates.length > 0) {
                accept('error', 'Some rules exported by this grammar are also included in other imports: ' + (0, stream_1.stream)(duplicates).distinct().join(', '), { node: imp, property: 'path' });
            }
        }
    }
    getDuplicateExportedRules(outer, inner) {
        const exclusiveOuter = outer.filter(g => !inner.includes(g));
        const outerRules = exclusiveOuter.flatMap(e => e.rules);
        const innerRules = inner.flatMap(e => e.rules);
        const duplicates = new Set();
        for (const outerRule of outerRules) {
            const outerName = outerRule.name;
            for (const innerRule of innerRules) {
                const innerName = innerRule.name;
                if (outerName === innerName) {
                    duplicates.add(innerRule.name);
                }
            }
        }
        return duplicates;
    }
    checkGrammarTypeInfer(grammar, accept) {
        var _a;
        const types = new Set();
        for (const type of grammar.types) {
            types.add(type.name);
        }
        for (const interfaceType of grammar.interfaces) {
            types.add(interfaceType.name);
        }
        for (const rule of grammar.rules.filter(ast.isParserRule)) {
            const isDataType = (0, grammar_util_1.isDataTypeRule)(rule);
            if (!isDataType && ((_a = rule.type) === null || _a === void 0 ? void 0 : _a.name) && types.has(rule.type.name) === !!rule.infers) {
                const keywordNode = rule.infers ? (0, grammar_util_1.findKeywordNode)(rule.$cstNode, 'infers') : (0, grammar_util_1.findKeywordNode)(rule.$cstNode, 'returns');
                accept('error', getMessage(rule.type.name, rule.infers), {
                    node: rule.type,
                    property: 'name',
                    code: rule.infers ? IssueCodes.InvalidInfers : IssueCodes.InvalidReturns,
                    data: keywordNode && (0, cst_util_1.toDocumentSegment)(keywordNode)
                });
            }
            else if (isDataType && rule.infers) {
                const inferNode = (0, grammar_util_1.findKeywordNode)(rule.$cstNode, 'infers');
                accept('error', 'Data type rules cannot infer a type.', {
                    node: rule,
                    property: 'infers',
                    code: IssueCodes.InvalidInfers,
                    data: inferNode && (0, cst_util_1.toDocumentSegment)(inferNode)
                });
            }
        }
        for (const action of (0, ast_util_1.streamAllContents)(grammar).filter(ast.isAction)) {
            if (action.type && types.has(action.type) === !!action.infer) {
                const keywordNode = action.infer ? (0, grammar_util_1.findKeywordNode)(action.$cstNode, 'infer') : (0, grammar_util_1.findKeywordNode)(action.$cstNode, '{');
                accept('error', getMessage(action.type, action.infer), {
                    node: action,
                    property: 'type',
                    code: action.infer ? IssueCodes.SuperfluousInfer : IssueCodes.MissingInfer,
                    data: keywordNode && (0, cst_util_1.toDocumentSegment)(keywordNode)
                });
            }
        }
        function getMessage(name, infer) {
            if (infer) {
                return `The type '${name}' is already explicitly declared and cannot be inferred.`;
            }
            else {
                return `The type '${name}' is not explicitly declared and must be inferred.`;
            }
        }
    }
    checkGrammarHiddenTokens(grammar, accept) {
        if (grammar.definesHiddenTokens) {
            accept('error', 'Hidden terminals are declared at the terminal definition.', { node: grammar, property: 'definesHiddenTokens', code: IssueCodes.HiddenGrammarTokens });
        }
    }
    checkHiddenTerminalRule(terminalRule, accept) {
        if (terminalRule.hidden && terminalRule.fragment) {
            accept('error', 'Cannot use terminal fragments as hidden tokens.', { node: terminalRule, property: 'hidden' });
        }
    }
    checkEmptyTerminalRule(terminalRule, accept) {
        try {
            const regex = (0, grammar_util_1.terminalRegex)(terminalRule);
            if (new RegExp(regex).test('')) {
                accept('error', 'This terminal could match an empty string.', { node: terminalRule, property: 'name' });
            }
        }
        catch (_a) {
            // In case the terminal can't be transformed into a regex, we throw an error
            // As this indicates unresolved cross references or parser errors, we can ignore this here
        }
    }
    checkUsedHiddenTerminalRule(ruleCall, accept) {
        const parentRule = (0, ast_util_1.getContainerOfType)(ruleCall, (n) => ast.isTerminalRule(n) || ast.isParserRule(n));
        if (parentRule) {
            if ('hidden' in parentRule && parentRule.hidden) {
                return;
            }
            const ref = ruleCall.rule.ref;
            if (ast.isTerminalRule(ref) && ref.hidden) {
                accept('error', 'Cannot use hidden terminal in non-hidden rule', { node: ruleCall, property: 'rule' });
            }
        }
    }
    checkUsedFragmentTerminalRule(ruleCall, accept) {
        const terminal = ruleCall.rule.ref;
        if (ast.isTerminalRule(terminal) && terminal.fragment) {
            const parentRule = (0, ast_util_1.getContainerOfType)(ruleCall, ast.isParserRule);
            if (parentRule) {
                accept('error', 'Cannot use terminal fragments as part of parser rules.', { node: ruleCall, property: 'rule' });
            }
        }
    }
    checkCrossReferenceSyntax(crossRef, accept) {
        if (crossRef.deprecatedSyntax) {
            accept('error', "'|' is deprecated. Please, use ':' instead.", { node: crossRef, property: 'deprecatedSyntax', code: IssueCodes.CrossRefTokenSyntax });
        }
    }
    checkPackageImport(imp, accept) {
        const resolvedGrammar = (0, grammar_util_1.resolveImport)(this.documents, imp);
        if (resolvedGrammar === undefined) {
            accept('error', 'Import cannot be resolved.', { node: imp, property: 'path' });
        }
        else if (imp.path.endsWith('.langium')) {
            accept('warning', 'Imports do not need file extensions.', { node: imp, property: 'path', code: IssueCodes.UnnecessaryFileExtension });
        }
    }
    checkGrammarImports(grammar, accept) {
        // Compute transitive grammar dependencies once for each grammar
        const importedGrammars = new Set((0, grammar_util_1.resolveTransitiveImports)(this.documents, grammar).map(e => (0, ast_util_1.getDocument)(e)));
        (0, ast_util_1.streamAllContents)(grammar).forEach(e => {
            if (ast.isRuleCall(e) || ast.isTerminalRuleCall(e)) {
                this.checkRuleCallImport(e, importedGrammars, accept);
            }
        });
    }
    checkRuleCallImport(ruleCall, importedDocuments, accept) {
        var _a;
        const ref = ruleCall.rule.ref;
        if (ref) {
            const refDoc = (0, ast_util_1.getDocument)(ref);
            const document = (0, ast_util_1.getDocument)(ruleCall);
            const grammar = document.parseResult.value;
            // Only check if the rule is sourced from another document
            if (ast.isGrammar(grammar) && refDoc !== document && !importedDocuments.has(refDoc)) {
                let relative = path_1.default.relative(vscode_uri_1.Utils.dirname(document.uri).fsPath, refDoc.uri.fsPath);
                if (relative.endsWith('.langium')) {
                    relative = relative.substring(0, relative.length - '.langium'.length);
                }
                if (!relative.startsWith('.')) {
                    relative = './' + relative;
                }
                accept('error', `Referenced rule "${(_a = ruleCall.rule.ref) === null || _a === void 0 ? void 0 : _a.name}" is not imported.`, {
                    node: ruleCall,
                    property: 'rule',
                    code: IssueCodes.MissingImport,
                    data: relative
                });
            }
        }
    }
    checkGrammarTypeAliases(grammar, accept) {
        var _a;
        const types = new Set();
        for (const type of grammar.types) {
            types.add(type.name);
        }
        for (const rule of grammar.rules) {
            if (((_a = rule.type) === null || _a === void 0 ? void 0 : _a.name) && types.has(rule.type.name)) {
                accept('error', 'Rules are not allowed to return alias types.', { node: rule.type, property: 'name' });
            }
        }
        for (const interfaceType of grammar.interfaces) {
            interfaceType.superTypes.forEach((superType, i) => {
                if (superType.ref && ast.isType(superType.ref)) {
                    accept('error', 'Interfaces cannot extend alias types.', { node: interfaceType, property: 'superTypes', index: i });
                }
            });
        }
        for (const action of (0, ast_util_1.streamAllContents)(grammar).filter(ast.isAction)) {
            if (action.type && types.has(action.type)) {
                accept('error', 'Actions cannot create alias types.', { node: action, property: 'type' });
            }
        }
    }
    checkTypesConsistency(grammar, accept) {
        (0, type_validator_1.validateTypesConsistency)(grammar, accept);
    }
    checkPropertyNameDuplication(grammar, accept) {
        var _a;
        if (grammar.interfaces.length === 0)
            return;
        const nameToInterfaceInfo = (0, type_validator_1.collectAllInterfaces)(grammar);
        for (const interfaceName of grammar.interfaces.map(e => e.name)) {
            const propertyNameToNode = new collections_1.MultiMap();
            this.collectPropertyNamesForHierarchy(nameToInterfaceInfo, new Set(), propertyNameToNode, interfaceName);
            for (const propertyName of propertyNameToNode.keys()) {
                const nodes = (_a = propertyNameToNode.get(propertyName)) !== null && _a !== void 0 ? _a : [];
                if (nodes.length < 2)
                    continue;
                for (const node of nodes) {
                    const errorMessage = `A property '${propertyName}' has to be unique for the whole hierarchy.`;
                    if (ast.isInterface(node)) {
                        const attributeNode = node.attributes.find(e => e.name === propertyName);
                        if (attributeNode) {
                            accept('error', errorMessage, { node: attributeNode, property: 'name' });
                        }
                    }
                    else {
                        (0, type_validator_1.applyErrorToAssignment)(node, accept)(propertyName, errorMessage);
                    }
                }
            }
        }
    }
    collectPropertyNamesForHierarchy(nameToInterfaceInfo, visited, result, interfaceName) {
        function collectPropertyNamesForHierarchyInternal(interfaceName) {
            if (visited.has(interfaceName))
                return;
            visited.add(interfaceName);
            const interfaceInfo = nameToInterfaceInfo.get(interfaceName);
            if (interfaceInfo) {
                interfaceInfo.type.properties.forEach(propery => result.add(propery.name, interfaceInfo.node));
                interfaceInfo.type.printingSuperTypes.forEach(superType => collectPropertyNamesForHierarchyInternal(superType));
            }
        }
        collectPropertyNamesForHierarchyInternal(interfaceName);
    }
    checkInvalidCharacterRange(range, accept) {
        if (range.right) {
            const message = 'Character ranges cannot use more than one character';
            let invalid = false;
            if (range.left.value.length > 1) {
                invalid = true;
                accept('error', message, { node: range.left, property: 'value' });
            }
            if (range.right.value.length > 1) {
                invalid = true;
                accept('error', message, { node: range.right, property: 'value' });
            }
            if (!invalid) {
                accept('hint', 'Consider using regex instead of character ranges', { node: range, code: IssueCodes.UseRegexTokens });
            }
        }
    }
    checkGrammarForUnusedRules(grammar, accept) {
        const visitedSet = new Set();
        const entry = (0, grammar_util_1.getEntryRule)(grammar);
        if (entry) {
            this.ruleDfs(entry, visitedSet);
            visitedSet.add(entry.name);
        }
        for (const rule of grammar.rules) {
            if (ast.isTerminalRule(rule) && rule.hidden) {
                continue;
            }
            if (!visitedSet.has(rule.name)) {
                accept('hint', 'This rule is declared but never referenced.', { node: rule, property: 'name', tags: [vscode_languageserver_types_1.DiagnosticTag.Unnecessary] });
            }
        }
    }
    ruleDfs(rule, visitedSet) {
        (0, ast_util_1.streamAllContents)(rule).forEach(node => {
            if (ast.isRuleCall(node)) {
                const refRule = node.rule.ref;
                if (refRule && !visitedSet.has(refRule.name)) {
                    visitedSet.add(refRule.name);
                    if (ast.isParserRule(refRule)) {
                        this.ruleDfs(refRule, visitedSet);
                    }
                }
            }
        });
    }
    checkRuleName(rule, accept) {
        if (rule.name) {
            const firstChar = rule.name.substring(0, 1);
            if (firstChar.toUpperCase() !== firstChar) {
                accept('warning', 'Rule name should start with an upper case letter.', { node: rule, property: 'name', code: IssueCodes.RuleNameUppercase });
            }
        }
    }
    checkKeyword(keyword, accept) {
        if (keyword.value.length === 0) {
            accept('error', 'Keywords cannot be empty.', { node: keyword });
        }
        else if (keyword.value.trim().length === 0) {
            accept('error', 'Keywords cannot only consist of whitespace characters.', { node: keyword });
        }
        else if (/\s/g.test(keyword.value)) {
            accept('warning', 'Keywords should not contain whitespace characters.', { node: keyword });
        }
    }
    checkUnorderedGroup(unorderedGroup, accept) {
        accept('error', 'Unordered groups are currently not supported', { node: unorderedGroup });
    }
    checkRuleParametersUsed(rule, accept) {
        const parameters = rule.parameters;
        if (parameters.length > 0) {
            const allReferences = (0, ast_util_1.streamAllContents)(rule).filter(ast.isParameterReference);
            for (const parameter of parameters) {
                if (!allReferences.some(e => e.parameter.ref === parameter)) {
                    accept('hint', `Parameter '${parameter.name}' is unused.`, {
                        node: parameter,
                        tags: [vscode_languageserver_types_1.DiagnosticTag.Unnecessary]
                    });
                }
            }
        }
    }
    checkParserRuleDataType(rule, accept) {
        var _a;
        const hasDatatypeReturnType = ((_a = rule.type) === null || _a === void 0 ? void 0 : _a.name) && isPrimitiveType(rule.type.name);
        const isDataType = (0, grammar_util_1.isDataTypeRule)(rule);
        if (!hasDatatypeReturnType && isDataType) {
            accept('error', 'This parser rule does not create an object. Add a primitive return type or an action to the start of the rule to force object instantiation.', { node: rule, property: 'name' });
        }
        else if (hasDatatypeReturnType && !isDataType) {
            accept('error', 'Normal parser rules are not allowed to return a primitive value. Use a datatype rule for that.', { node: rule.type, property: 'name' });
        }
    }
    checkTerminalRuleReturnType(rule, accept) {
        var _a;
        if (((_a = rule.type) === null || _a === void 0 ? void 0 : _a.name) && !isPrimitiveType(rule.type.name)) {
            accept('error', "Terminal rules can only return primitive types like 'string', 'boolean', 'number' or 'date'.", { node: rule.type, property: 'name' });
        }
    }
    checkRuleCallParameters(ruleCall, accept) {
        const rule = ruleCall.rule.ref;
        if (ast.isParserRule(rule)) {
            const expected = rule.parameters.length;
            const given = ruleCall.arguments.length;
            if (expected !== given) {
                accept('error', `Rule '${rule.name}' expects ${expected} arguments, but got ${given}.`, { node: ruleCall });
            }
        }
        else if (ast.isTerminalRule(rule) && ruleCall.arguments.length > 0) {
            accept('error', 'Terminal rules do not accept any arguments', { node: ruleCall });
        }
    }
    checkCrossRefNameAssignment(reference, accept) {
        if (!reference.terminal && reference.type.ref && !(0, grammar_util_1.findNameAssignment)(reference.type.ref)) {
            accept('error', 'Cannot infer terminal or data type rule for cross reference.', { node: reference, property: 'type' });
        }
    }
    checkCrossRefTerminalType(reference, accept) {
        if (ast.isRuleCall(reference.terminal) && ast.isParserRule(reference.terminal.rule.ref) && !(0, grammar_util_1.isDataTypeRule)(reference.terminal.rule.ref)) {
            accept('error', 'Parser rules cannot be used for cross references.', { node: reference.terminal, property: 'rule' });
        }
    }
    checkCrossRefType(reference, accept) {
        const issue = this.checkReferenceToRuleButNotType(reference.type);
        if (issue) {
            accept('error', issue, { node: reference, property: 'type' });
        }
    }
    checkAtomTypeRefType(atomType, accept) {
        const issue = this.checkReferenceToRuleButNotType(atomType.refType);
        if (issue) {
            accept('error', issue, { node: atomType, property: 'refType' });
        }
    }
    checkReferenceToRuleButNotType(type) {
        if (type && ast.isParserRule(type.ref) && !(0, grammar_util_1.isDataTypeRule)(type.ref) && type.ref.type) {
            return `Use the rule type '${type.ref.type.name}' instead of the typed rule '${type.ref.name}' for cross references.`;
        }
        return undefined;
    }
}
exports.LangiumGrammarValidator = LangiumGrammarValidator;
const primitiveTypes = ['string', 'number', 'boolean', 'Date'];
function isPrimitiveType(type) {
    return primitiveTypes.includes(type);
}
