// langium/lib/grammar/type-system/inferred-types.js
"use strict";
/******************************************************************************
 * Copyright 2022 TypeFox GmbH
 * This program and the accompanying materials are made available under the
 * terms of the MIT License, which is available in the project root.
 ******************************************************************************/
Object.defineProperty(exports, "__esModule", { value: true });
exports.collectInferredTypes = void 0;
const grammar_util_1 = require("../grammar-util");
const ast_1 = require("../generated/ast");
const stream_1 = require("../../utils/stream");
const types_util_1 = require("./types-util");
class TypeTree {
    constructor(root) {
        this.descendents = new Map();
        this.descendents.set(root, []);
    }
    split(type, count) {
        const descendents = new Array(count).fill(JSON.parse(JSON.stringify(type)));
        descendents.forEach(e => this.descendents.set(e, []));
        this.descendents.set(type, descendents);
        return descendents;
    }
    getLeafNodesOf(type) {
        var _a;
        const leaves = [];
        const direct = (_a = this.descendents.get(type)) !== null && _a !== void 0 ? _a : [];
        for (const directDesc of direct) {
            if (!this.hasLeaves(directDesc)) {
                leaves.push(directDesc);
            }
            else {
                leaves.push(...this.getLeafNodesOf(directDesc));
            }
        }
        if (leaves.length === 0) {
            leaves.push(type);
        }
        return leaves;
    }
    getLeafNodes() {
        return Array.from(this.descendents.keys());
    }
    hasLeaves(type) {
        var _a;
        return ((_a = this.descendents.get(type)) !== null && _a !== void 0 ? _a : []).length > 0;
    }
}
class TypeCollectorState {
    constructor(type) {
        this.cardinalities = [];
        this.tree = new TypeTree(type);
    }
    enterGroup(cardinality) {
        this.cardinalities.push(cardinality);
    }
    leaveGroup() {
        this.cardinalities.pop();
    }
    isOptional() {
        return this.cardinalities.some(e => (0, grammar_util_1.isOptional)(e));
    }
}
function collectInferredTypes(parserRules, datatypeRules) {
    var _a, _b;
    // extract interfaces and types from parser rules
    const allTypes = [];
    for (const rule of parserRules) {
        const type = initType(rule);
        const state = new TypeCollectorState(type);
        state.parserRule = rule;
        collectElement(state, type, rule.alternatives);
        allTypes.push(...state.tree.getLeafNodes());
    }
    const interfaces = calculateAst(allTypes);
    buildContainerTypes(interfaces);
    const inferredTypes = extractTypes(interfaces);
    // extract types from datatype rules
    for (const rule of datatypeRules) {
        const types = (0, ast_1.isAlternatives)(rule.alternatives) && rule.alternatives.elements.every(e => (0, ast_1.isKeyword)(e)) ?
            (0, stream_1.stream)(rule.alternatives.elements).filter(ast_1.isKeyword).map(e => `'${e.value}'`).toArray() :
            [(_b = (_a = rule.type) === null || _a === void 0 ? void 0 : _a.name) !== null && _b !== void 0 ? _b : 'string'];
        inferredTypes.types.push(new types_util_1.TypeType(rule.name, [{ types, reference: false, array: false }]));
    }
    return inferredTypes;
}
exports.collectInferredTypes = collectInferredTypes;
// todo simplify
function initType(rule) {
    return {
        name: (0, ast_1.isParserRule)(rule) ? (0, grammar_util_1.getTypeName)(rule) : rule,
        super: [],
        properties: [],
        ruleCalls: [],
        hasAction: false
    };
}
/**
 * Collects all possible type branches of a given element.
 * @param state State to walk over element's graph.
 * @param type Element that collects a current type branch for the given element.
 * @param element The given AST element, from which it's necessary to extract the type.
 */
function collectElement(state, type, element) {
    state.currentType = type;
    state.enterGroup(element.cardinality);
    if ((0, grammar_util_1.isOptional)(element.cardinality)) {
        state.tree.split(state.currentType, 2);
    }
    if ((0, ast_1.isAlternatives)(element)) {
        state.tree
            .split(state.currentType, element.elements.length)
            .forEach((split, i) => collectElement(state, split, element.elements[i]));
    }
    else if ((0, ast_1.isGroup)(element) || (0, ast_1.isUnorderedGroup)(element)) {
        for (const item of element.elements) {
            state.tree
                .getLeafNodesOf(type)
                .forEach(leaf => collectElement(state, leaf, item));
        }
    }
    else if ((0, ast_1.isAction)(element)) {
        addAction(state, element);
    }
    else if ((0, ast_1.isAssignment)(element)) {
        addAssignment(state, element);
    }
    else if ((0, ast_1.isRuleCall)(element)) {
        addRuleCall(state, element);
    }
    state.leaveGroup();
}
function addAction(state, action) {
    const type = state.currentType;
    if (type) {
        let newType;
        if (action.type !== type.name) {
            newType = initType(action.type);
            newType.super.push((0, grammar_util_1.getTypeName)(state.parserRule));
            newType.hasAction = true;
            state.tree.descendents.get(type).push(newType);
            state.tree.descendents.set(newType, []);
            state.currentType = newType;
        }
        else {
            newType = type;
        }
        if (action.feature && action.operator) {
            newType.properties.push({
                name: action.feature,
                optional: false,
                typeAlternatives: [{
                        array: action.operator === '+=',
                        reference: false,
                        types: [(0, grammar_util_1.getTypeName)(state.parserRule)]
                    }]
            });
        }
    }
}
function addAssignment(state, assignment) {
    var _a;
    const typeItems = { types: [], reference: false };
    findTypes(assignment.terminal, typeItems);
    (_a = state.currentType) === null || _a === void 0 ? void 0 : _a.properties.push({
        name: assignment.feature,
        optional: (0, grammar_util_1.isOptional)(assignment.cardinality) || state.isOptional(),
        typeAlternatives: [{
                array: assignment.operator === '+=',
                types: assignment.operator === '?=' ? ['boolean'] : Array.from(new Set(typeItems.types)),
                reference: typeItems.reference
            }]
    });
}
function findTypes(terminal, types) {
    if ((0, ast_1.isAlternatives)(terminal) || (0, ast_1.isUnorderedGroup)(terminal) || (0, ast_1.isGroup)(terminal)) {
        findInCollection(terminal, types);
    }
    else if ((0, ast_1.isKeyword)(terminal)) {
        types.types.push(`'${terminal.value}'`);
    }
    else if ((0, ast_1.isRuleCall)(terminal) && terminal.rule.ref) {
        types.types.push((0, grammar_util_1.getRuleType)(terminal.rule.ref));
    }
    else if ((0, ast_1.isCrossReference)(terminal) && terminal.type.ref) {
        types.types.push((0, grammar_util_1.getTypeName)(terminal.type.ref));
        types.reference = true;
    }
}
function findInCollection(collection, types) {
    for (const element of collection.elements) {
        findTypes(element, types);
    }
}
function addRuleCall(state, ruleCall) {
    const rule = ruleCall.rule.ref;
    const type = state.currentType;
    if (type) {
        // Add all properties of fragments to the current type
        if ((0, ast_1.isParserRule)(rule) && rule.fragment) {
            const fragmentType = initType(rule);
            const fragmentState = new TypeCollectorState(fragmentType);
            collectElement(fragmentState, fragmentType, rule.alternatives);
            const types = calculateAst(fragmentState.tree.getLeafNodes());
            const foundType = types.find(e => e.name === rule.name);
            if (foundType) {
                type.properties.push(...foundType.properties);
            }
        }
        else if ((0, ast_1.isParserRule)(rule)) {
            type.ruleCalls.push((0, grammar_util_1.getRuleType)(rule));
        }
    }
}
/**
 * Calculate interfaces from all possible type branches.
 * [some of these interfaces will become types in the generated AST]
 * @param alternatives The type branches that will be squashed in interfaces.
 * @returns Interfaces.
 */
function calculateAst(alternatives) {
    const interfaces = [];
    const ruleCallAlternatives = [];
    const flattened = flattenTypes(alternatives);
    for (const flat of flattened) {
        const interfaceType = new types_util_1.InterfaceType(flat.name, flat.super, flat.properties);
        interfaces.push(interfaceType);
        if (flat.ruleCalls.length > 0) {
            ruleCallAlternatives.push(flat);
        }
        // all other cases assume we have a data type rule
        // we do not generate an AST type for data type rules
    }
    for (const ruleCallType of ruleCallAlternatives) {
        let exists = false;
        for (const ruleCall of ruleCallType.ruleCalls) {
            const calledInterface = interfaces.find(e => e.name === ruleCall);
            if (calledInterface) {
                if (calledInterface.name === ruleCallType.name) {
                    exists = true;
                }
                else {
                    calledInterface.superTypes.push(ruleCallType.name);
                }
            }
        }
        if (!exists && !interfaces.some(e => e.name === ruleCallType.name)) {
            interfaces.push(new types_util_1.InterfaceType(ruleCallType.name, ruleCallType.super, []));
        }
    }
    for (const interfaceType of interfaces) {
        interfaceType.superTypes = Array.from(new Set(interfaceType.superTypes));
    }
    return interfaces;
}
function flattenTypes(alternatives) {
    const names = new Set(alternatives.map(e => e.name));
    const types = [];
    for (const name of names) {
        const properties = [];
        const ruleCalls = new Set();
        const type = { name, properties, ruleCalls: [], super: [], hasAction: false };
        const namedAlternatives = alternatives.filter(e => e.name === name);
        for (const alt of namedAlternatives) {
            type.super.push(...alt.super);
            type.hasAction = type.hasAction || alt.hasAction;
            const altProperties = alt.properties;
            for (const altProperty of altProperties) {
                const existingProperty = properties.find(e => e.name === altProperty.name);
                if (existingProperty) {
                    existingProperty.optional = existingProperty.optional && altProperty.optional;
                    altProperty.typeAlternatives.filter(isNotInTypeAlternatives(existingProperty.typeAlternatives)).forEach(type => existingProperty.typeAlternatives.push(type));
                }
                else {
                    properties.push(Object.assign({}, altProperty));
                }
            }
            if (altProperties.length === 0) {
                alt.ruleCalls.forEach(ruleCall => ruleCalls.add(ruleCall));
            }
        }
        type.ruleCalls = Array.from(ruleCalls);
        types.push(type);
    }
    return types;
}
function isNotInTypeAlternatives(typeAlternatives) {
    return (type) => {
        return !typeAlternatives.some(e => comparePropertyType(e, type));
    };
}
function comparePropertyType(a, b) {
    return a.array === b.array &&
        a.reference === b.reference &&
        compareLists(a.types, b.types);
}
function compareLists(a, b, eq = (x, y) => x === y) {
    if (a.length !== b.length)
        return false;
    const distictAndSortedA = (0, types_util_1.distictAndSorted)(a);
    return (0, types_util_1.distictAndSorted)(b).every((e, i) => eq(e, distictAndSortedA[i]));
}
/**
 * Builds container types for given interfaces.
 * @param interfaces The interfaces that have to get container types.
 */
function buildContainerTypes(interfaces) {
    var _a, _b;
    // 1st stage: collect container types & calculate sub-types
    for (const interfaceType of interfaces) {
        for (const typeName of interfaceType.properties.flatMap(property => property.typeAlternatives.filter(e => !e.reference).flatMap(e => e.types))) {
            (_a = interfaces.find(e => e.name === typeName)) === null || _a === void 0 ? void 0 : _a.containerTypes.push(interfaceType.name);
        }
        for (const superTypeName of interfaceType.superTypes) {
            (_b = interfaces.find(e => e.name === superTypeName)) === null || _b === void 0 ? void 0 : _b.subTypes.push(interfaceType.name);
        }
    }
    // 2nd stage: share container types
    const connectedComponents = [];
    calculateConnectedComponents(connectedComponents, interfaces);
    shareContainerTypes(connectedComponents);
}
function calculateConnectedComponents(connectedComponents, interfaces) {
    function dfs(typeInterface) {
        const component = [typeInterface];
        visited.add(typeInterface.name);
        for (const nextTypeInterfaceName of typeInterface.subTypes.concat(typeInterface.superTypes)) {
            if (!visited.has(nextTypeInterfaceName)) {
                const nextTypeInterface = interfaces.find(e => e.name === nextTypeInterfaceName);
                if (nextTypeInterface) {
                    component.push(...dfs(nextTypeInterface));
                }
            }
        }
        return component;
    }
    const visited = new Set();
    for (const typeInterface of interfaces) {
        if (!visited.has(typeInterface.name)) {
            connectedComponents.push(dfs(typeInterface));
        }
    }
}
function shareContainerTypes(connectedComponents) {
    for (const component of connectedComponents) {
        const containerTypes = [];
        component.forEach(type => containerTypes.push(...type.containerTypes));
        component.forEach(type => type.containerTypes = containerTypes);
    }
}
/**
 * Filters interfaces, transforming some of them in types.
 * The transformation criterion: no properties, but have subtypes.
 * @param interfaces The interfaces that have to be transformed on demand.
 * @returns Types and not transformed interfaces.
 */
function extractTypes(interfaces) {
    const astTypes = { interfaces: [], types: [] };
    const typeNames = [];
    for (const interfaceType of interfaces) {
        // the criterion for converting an interface into a type
        if (interfaceType.properties.length === 0 && interfaceType.subTypes.length > 0) {
            const alternatives = interfaceType.subTypes.map(e => ({
                types: [e],
                reference: false,
                array: false
            }));
            const type = new types_util_1.TypeType(interfaceType.name, alternatives, { reflection: true });
            type.superTypes = interfaceType.superTypes;
            astTypes.types.push(type);
            typeNames.push(interfaceType.name);
        }
        else {
            astTypes.interfaces.push(interfaceType);
        }
    }
    // define printingSuperTypes containing intefaces that
    // became types from super types of their "former" children
    for (const interfaceType of astTypes.interfaces) {
        interfaceType.printingSuperTypes = JSON.parse(JSON.stringify(interfaceType.superTypes))
            .filter(superType => !typeNames.includes(superType));
    }
    return astTypes;
}
